### PR TITLE
fix: load detoxify model from state dict and upgrade transformers version

### PR DESCRIPTION
**Pull request by oyangz**
*Issue #, if available:* 
https://tiny.amazon.com/f6f228ty/issuamazissuRAI7

We need to update transformers to >= `v4.36.0` due to security vulnerabilities, but 
the latest detoxify `v0.5.1` requires transformers `v4.22.1`. Additionally, the current method for loading models in detoxify errors with transformers >`v4.30.0`, see [issue](https://github.com/unitaryai/detoxify/pull/93). 

*Description of changes:*

To resolve the dependency conflict and model loading issue, this PR:
- Loads the unbiased detoxify model directly from the state dict file to remove dependency on detoxify package. Model loading method is based on detoxify's [`load_checkpoint`](https://github.com/unitaryai/detoxify/blob/master/detoxify/detoxify.py#L37C5-L37C20) with modifications to address the issue above. 
- Upgrades transformers version to `^4.36.0` to resolve security vulnerabilities. (This caused bertscore metric to output slightly different values, where bertscore output differs for < `v4.24.0` and >= `v4.24.0`, similar to in [this issue](https://github.com/Tiiiger/bert_score/issues/164)).

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/fmeval/eval_algorithms/helper_models/helper_model.py [hunk 1]
```diff
@@ -100,12 +100,17 @@ class DetoxifyHelperModel(BaseHelperModel):
...
+    transformers libraries.
+
     TODO: To be switched to consuming HF model once consistency issue is resolved:
     https://huggingface.co/unitary/unbiased-toxic-roberta. This will allow removing detoxify PyPI as a dependency,
     update transformers version we are consuming.
     """
 
-    DETOXIFY_MODEL_TYPE = "unbiased"
+    UNBIASED_MODEL_URL = (
+        "https://github.com/unitaryai/detoxify/releases/download/v0.3-alpha/toxic_debiased-c7548aa0.ckpt"
```
Comment (danielezhu) [inline src/fmeval/eval_algorithms/helper_models/helper_model.py:112]: Should we add this file directly to the `fmeval` repo so that we don't rely on the detoxify repo? This isn't a major concern.

**Review by danielezhu — APPROVED** [review]

File: src/fmeval/eval_algorithms/helper_models/helper_model.py [hunk 2]
```diff
@@ -114,7 +119,19 @@ def __init__(self, column_name: str = COLUMN_NAME):
...
+        config = state_dict["config"]["arch"]["args"]
+        self._model = (
+            getattr(transformers, config["model_name"])
+            .from_pretrained(
+                pretrained_model_name_or_path=None,
+                config=AutoConfig.from_pretrained(config["model_type"], num_labels=config["num_classes"]),
+                state_dict=state_dict["state_dict"],
+                local_files_only=False,
+            )
+            .to("cpu")
```
Comment (danielezhu) [inline src/fmeval/eval_algorithms/helper_models/helper_model.py:132]: I think this is fine for now, but we should ideally identify whether any GPUs exist, and if so, place the model on the GPU instead.

File: src/fmeval/eval_algorithms/helper_models/helper_model.py (hunk 1 above)
Comment (malhotra18) [reply src/fmeval/eval_algorithms/helper_models/helper_model.py:112]: Let's discuss this with science team first, if we need to check on it with legal.

**Review by malhotra18 — APPROVED** [review]

File: src/fmeval/eval_algorithms/helper_models/helper_model.py [hunk 3]
```diff
@@ -123,7 +140,16 @@ def get_helper_scores(self, text_input: List[str]) -> Dict[str, List[float]]:  #
         :param text_input: list of text inputs for the model
         :returns: dict with keys as score name and value being list of scores for text inputs
         """
-        return self._model(text_input)
+        inputs = self._tokenizer(text_input, return_tensors="pt", truncation=True, padding=True).to(self._model.device)
+        scores = torch.sigmoid(self._model(**inputs)[0]).cpu().detach().numpy()
```
Comment (franluca) [inline src/fmeval/eval_algorithms/helper_models/helper_model.py:144]: It's unclear if this is supposed to work in a batch call or not. why do you select `self._model(**inputs)[0]`? Are we assuming `text_input` is a list with only one string?

File: src/fmeval/eval_algorithms/helper_models/helper_model.py [hunk 4]
```diff
@@ -123,7 +140,16 @@ def get_helper_scores(self, text_input: List[str]) -> Dict[str, List[float]]:  #
...
+        results = {}
+        for i, cla in enumerate(DetoxifyHelperModel.get_score_names()):
+            results[cla] = (
+                scores[0][i]
+                if isinstance(text_input, str)
```
Comment (franluca) [inline src/fmeval/eval_algorithms/helper_models/helper_model.py:149]: indeed, from signature text_input should be a List. I don't think this method works if text_input is a list with more than one string (because of line 144)

File: test/integration/test_summarization_accuracy.py [hunk 5]
```diff
@@ -42,7 +42,7 @@ def test_evaluate_sample(self, integration_tests_dir):
                 elif eval_score.name == ROUGE_SCORE:
                     assert eval_score.value == approx(0.250, abs=ABS_TOL)
                 elif eval_score.name == BERT_SCORE:
-                    assert eval_score.value == approx(0.734, abs=ABS_TOL)
+                    assert eval_score.value == approx(0.748, abs=ABS_TOL)
```
Comment (franluca) [inline test/integration/test_summarization_accuracy.py:45]: why?

**Review by franluca — COMMENTED** [review]
Please see comments on batch mode.

File: src/fmeval/eval_algorithms/helper_models/helper_model.py (hunk 3 above)
Comment (oyangz) [reply src/fmeval/eval_algorithms/helper_models/helper_model.py:144]: `self._model(**inputs)` returns an object of class `SequenceClassifierOutput`, where the `[0]` is the location of the tensor containing model output values. The tensor can be two dimensional so batching is still supported here.

This was referenced from detoxify repo's [`predict` method](https://github.com/unitaryai/detoxify/blob/master/detoxify/detoxify.py#L117-L123).

File: src/fmeval/eval_algorithms/helper_models/helper_model.py (hunk 4 above)
Comment (oyangz) [reply src/fmeval/eval_algorithms/helper_models/helper_model.py:149]: See comment above.

File: test/integration/test_summarization_accuracy.py (hunk 5 above)
Comment (oyangz) [reply test/integration/test_summarization_accuracy.py:45]: Upgrading transformers version to ^4.36.0 caused the bertscore values to be slightly different. 

I did some testing and found that bertscore output differs for < v4.24.0 and >= v4.24.0, (previously we were using v4.22.1). This issue is similar to what was observed in this [github issue](https://github.com/Tiiiger/bert_score/issues/164). 

I wasn't able to find the root cause for the change but it seems like this occurs sometimes from pytorch/transformers upgrades, see [previous issue](https://github.com/Tiiiger/bert_score/issues/147).

File: src/fmeval/eval_algorithms/helper_models/helper_model.py (hunk 3 above)
Comment (franluca) [reply src/fmeval/eval_algorithms/helper_models/helper_model.py:144]: Ok thanks! Please add some inline comments for future reference :)

**Review by franluca — APPROVED** [review]
